File: src/test/java/io/github/mschout/email/srs/SRSTest.java
package io.github.mschout.email.srs;

import static org.junit.jupiter.api.Assertions.*;

import com.google.common.collect.ImmutableList;
import io.github.mschout.email.srs.provider.SRSProvider;
import io.github.mschout.email.srs.provider.SRSProviderFactory;
import java.security.InvalidKeyException;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import org.junit.jupiter.api.Test;

class SRSTest {

  @Test
  void guardedSRS() throws InvalidKeyException {
    SRS srs = SRS.guardedSRS(ImmutableList.of("dummy-secret"));

    String origAddress = "jdoe@example.com";

    String forward = srs.forward(origAddress, "other.com");
    assertNotEquals(origAddress, forward);

    String reverse = srs.reverse(forward);
    assertEquals(reverse, origAddress);
  }

  @Test
  void isSRS() {
    SRS srs = SRS.guardedSRS(ImmutableList.of("dummy-secret"));

    assertTrue(srs.isSRS("SRS0=5gnp=ZU=example.com=jdoe@other.com"));
    assertFalse(srs.isSRS("jdoe@example.com"));
  }

  @Test
  public void separatorTests() throws InvalidKeyException {
    List<String> tests = ImmutableList.of(
      "user@domain-with-dash.com",
      "user-with-dash@domain.com",
      "user+with+plus@domain.com",
      "user=with=equals@domain.com",
      "user%with!everything&everything=@domain.somewhere"
    );

    for (String separator : ImmutableList.of("-", "+", "=")) {
      SRSProvider provider = SRSProviderFactory
        .builder()
        .separator(separator)
        .hashMinLength(4)
        .hashLength(4)
        .build()
        .createProvider(SRS.Type.GUARDED, ImmutableList.of("foo"));

      SRS srs = new SRS(provider);

      assertInstanceOf(SRS.class, srs, "Created an object");
      assertEquals(separator, provider.getSeparator(), "Got the expected separator");

      String source = "user@host.tld";
      List<String> aliases = IntStream
        .rangeClosed(0, 5)
        .boxed()
        .map(i -> String.format("alias%s@host%s.tld%s", i, i, i))
        .collect(Collectors.toList());

      String srs0 = srs.forward(source, aliases.get(0));
      assertTrue(srs0.startsWith("SRS0" + separator), "It uses the right initial prefix");

      String old0 = srs.reverse(srs0);
      assertFalse(old0.isEmpty(), "Reversed the address");
      assertEquals(old0, source, "The reversal was idempotent");

      String srs1 = srs.forward(srs0, aliases.get(1));
      assertFalse(srs1.isEmpty(), "Made another new address with the SRS address");
      assertTrue(srs1.startsWith("SRS1" + separator), "It uses the right initial separator too");

      String old1 = srs.reverse(srs1);
      assertFalse(old1.isEmpty(), "Reversed the address again");
      assertTrue(old1.startsWith("SRS0" + separator), "Got an SRS0 address");
      assertEquals(srs0, old1, "It is the original SRS0 address");

      String orig = srs.reverse(old1);
      assertEquals(source, orig, "Got back the original sender");

      for (String test : tests) {
        String srs0Addr = srs.forward(test, aliases.get(0));
        String oldAddr = srs.reverse(srs0Addr);
        assertEquals(test, oldAddr, "Idempotent on " + test);

        String srs1Addr = srs.forward(srs0Addr, aliases.get(1));
        String srs0Rev = srs.reverse(srs1Addr);
        assertEquals(srs0Addr, srs0Rev, "Idempotent on " + srs0Addr);
      }
    }

    assertThrows(
      IllegalArgumentException.class,
      () ->
        SRSProviderFactory
          .builder()
          .separator("!")
          .hashMinLength(4)
          .hashLength(4)
          .build()
          .createProvider(SRS.Type.GUARDED, ImmutableList.of("foo")),
      "Failed to create an object with bad separator"
    );
  }
}

File: src/test/kotlin/io/github/mschout/email/srs/SRSTest.kt
/*
 * Copyright 2026 Michael Schout
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.mschout.email.srs

import io.github.mschout.email.srs.provider.SRSProviderFactory
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.booleans.shouldBeFalse
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldNotBeEmpty
import io.kotest.matchers.string.shouldStartWith
import io.kotest.matchers.types.shouldBeInstanceOf

class SRSTest :
    FunSpec({
      test("guardedSRS") {
        val srs = SRS.guardedSRS(listOf("dummy-secret"))

        val origAddress = "jdoe@example.com"

        val forward = srs.forward(origAddress, "other.com")
        forward shouldNotBe origAddress

        val reverse = srs.reverse(forward)
        reverse shouldBe origAddress
      }

      test("isSRS") {
        val srs = SRS.guardedSRS(listOf("dummy-secret"))

        srs.isSRS("SRS0=5gnp=ZU=example.com=jdoe@other.com").shouldBeTrue()
        srs.isSRS("jdoe@example.com").shouldBeFalse()
      }

      test("separatorTests") {
        val tests =
            listOf(
                "user@domain-with-dash.com",
                "user-with-dash@domain.com",
                "user+with+plus@domain.com",
                "user=with=equals@domain.com",
                "user%with!everything&everything=@domain.somewhere",
            )

        for (separator in listOf("-", "+", "=")) {
          val provider =
              SRSProviderFactory.builder()
                  .separator(separator)
                  .hashMinLength(4)
                  .hashLength(4)
                  .build()
                  .createProvider(SRS.Type.GUARDED, listOf("foo"))

          val srs = SRS(provider)

          srs.shouldBeInstanceOf<SRS>()
          provider.separator shouldBe separator

          val source = "user@host.tld"
          val aliases = (0..5).map { i -> "alias$i@host$i.tld$i" }

          val srs0 = srs.forward(source, aliases[0])
          srs0 shouldStartWith "SRS0$separator"

          val old0 = srs.reverse(srs0)
          old0.shouldNotBeEmpty()
          old0 shouldBe source

          val srs1 = srs.forward(srs0, aliases[1])
          srs1.shouldNotBeEmpty()
          srs1 shouldStartWith "SRS1$separator"

          val old1 = srs.reverse(srs1)
          old1.shouldNotBeEmpty()
          old1 shouldStartWith "SRS0$separator"
          srs0 shouldBe old1

          val orig = srs.reverse(old1)
          source shouldBe orig

          for (test in tests) {
            val srs0Addr = srs.forward(test, aliases[0])
            val oldAddr = srs.reverse(srs0Addr)
            test shouldBe oldAddr

            val srs1Addr = srs.forward(srs0Addr, aliases[1])
            val srs0Rev = srs.reverse(srs1Addr)
            srs0Addr shouldBe srs0Rev
          }
        }

        shouldThrow<IllegalArgumentException> {
          SRSProviderFactory.builder()
              .separator("!")
              .hashMinLength(4)
              .hashLength(4)
              .build()
              .createProvider(SRS.Type.GUARDED, listOf("foo"))
        }
      }

      test("varySeparator") {
        val tests =
            listOf(
                "user@domain-with-dash.com",
                "user-with-dash@domain.com",
                "user+with+plus@domain.com",
                "user=with=equals@domain.com",
                "user%with!everything&everything=@domain.somewhere",
            )

        val alias0 = "alias@host.com"
        val alias1 = "name@forwarder.com"
        val alias2 = "user@postal.com"

        for (type in listOf(SRS.Type.GUARDED, SRS.Type.REVERSIBLE, SRS.Type.SHORTCUT)) {
          val srs0 =
              SRS(
                  SRSProviderFactory.builder()
                      .separator("+")
                      .build()
                      .createProvider(type, listOf("foo"))
              )

          val srs1 =
              SRS(
                  SRSProviderFactory.builder()
                      .separator("-")
                      .build()
                      .createProvider(type, listOf("foo"))
              )

          val srs2 =
              SRS(
                  SRSProviderFactory.builder()
                      .separator("=")
                      .build()
                      .createProvider(type, listOf("foo"))
              )

          for (test in tests) {
            val srs0Addr = srs0.forward(test, alias0)
            val srs0Rev = srs0.reverse(srs0Addr)
            test shouldBe srs0Rev

            val srs1Addr = srs1.forward(srs0Addr, alias1)
            val srs1Rev = srs1.reverse(srs1Addr)

            if (type == SRS.Type.SHORTCUT) {
              test shouldBe srs1Rev
            } else {
              srs0Addr shouldBe srs1Rev
            }

            val srs2Addr = srs2.forward(srs1Addr, alias2)
            val srs2Rev = srs2.reverse(srs2Addr)

            if (type == SRS.Type.GUARDED) {
              srs0Addr shouldBe srs2Rev
            } else if (type == SRS.Type.REVERSIBLE) {
              srs1Addr shouldBe srs2Rev
            }
          }
        }
      }

      test("caselessTests") {
        for (type in listOf(SRS.Type.GUARDED, SRS.Type.REVERSIBLE, SRS.Type.SHORTCUT)) {
          val provider =
              SRSProviderFactory.builder()
                  .separator("+")
                  .build()
                  .createProvider(type, listOf("foo"))

          val srs = SRS(provider)

          val tests =
              listOf(
                  "User@domain-with-dash.com",
                  "User-with-dash@domain.com",
                  "User+with+plus@domain.com",
                  "User=with=equals@domain.com",
                  "User%with!everything&everything=@domain.somewhere",
              )

          val alias0 = "alias@host.com"
          val alias1 = "name@forwarder.com"

          for (test in tests) {
            val srs0Addr = srs.forward(test, alias0).lowercase()
            val srs0Rev = srs.reverse(srs0Addr)
            srs0Rev.equals(test, ignoreCase = true).shouldBeTrue()

            val srs1Addr = srs.forward(srs0Addr, alias1).lowercase()
            val srs1Rev = srs.reverse(srs1Addr)

            if (type == SRS.Type.SHORTCUT) {
              test.equals(srs1Rev, ignoreCase = true).shouldBeTrue()
            } else {
              srs0Addr.equals(srs1Rev, ignoreCase = true).shouldBeTrue()
            }
          }
        }
      }
    })
